Features/first relationship approach

kiuby88 opened the pull request:
First relations support.
A simple type is aded in order to configure any Entity ConfigKey.

I hope this PR will be a start point to discuss about relations management. 
I think @ahgittin thought like [that](https://gist.github.com/kiuby88/511a6d1fc185e452e08c).

sjcorbett commented on brooklyn-tosca-transformer/src/main/java/org/apache/brooklyn/tosca/a4c/brooklyn/ToscaNodeToEntityConverter.java:
```diff
@@ -146,17 +149,71 @@ public ToscaNodeToEntityConverter setNodeId(String nodeId) {
+                }
+                if(simpleProperty==null) {
+                    propertiesAndTypedValues=
+                            ImmutableMap.of(propCollection, ((Object)ImmutableList.of(propValue)));
```
`prop.collection` is not marked as required in the example `relationship.yaml` so this (and the case a few lines below)  could plausibly cause a NullPointerException.

sjcorbett commented on brooklyn-tosca-transformer/src/test/java/org/apache/brooklyn/tosca/a4c/brooklyn/ToscaPlanToSpecTransformerTest.java:
```diff
@@ -59,7 +65,7 @@ public void testSimpleHostedTopologyParser() {
     }
 
     // FIXME: Test fails when asserting the size of the tomcat server's config map.
-    @Test(enabled = false)
```
This test still fails.

kiuby88 replied:
Sorry. But, this test is running in my machine. What error did you find? I will try to solve it.

sjcorbett replied:
It's not failing for me now - I guess it was fixed by 4bd180d.

sjcorbett commented on brooklyn-tosca-transformer/src/test/java/org/apache/brooklyn/tosca/a4c/brooklyn/ToscaPlanToSpecTransformerTest.java:
```diff
@@ -70,21 +76,35 @@ public void testDslInChatApplication() {
+        assertEquals(app.getChildren().size(), 2);
+
+        EntitySpec<TomcatServer> tomcatServer = (EntitySpec<TomcatServer>) app.getChildren().get(0);
+        assertNotNull(tomcatServer.getConfig().get(TomcatServer.JAVA_SYSPROPS));
```
This assertion fails. If app has two children how do you know that `app.getChildren().get(0)` will give you the spec for the Tomcat server?

kiuby88 replied:
@sjcorbett I have added a [predicate](https://github.com/cloudsoft/brooklyn-tosca/pull/21/files#diff-70e0a598745438ee1b447d4c026179b9R37) to solving this issue. WDYT?

sjcorbett commented on brooklyn-tosca-transformer/src/test/resources/templates/relationship.yaml:
```diff
@@ -0,0 +1,90 @@
+            properties:
+              prop.collection: java.sysprops
+              prop.name: brooklyn.example.db.url
+              prop.value: $brooklyn:formatString("jdbc:%s%s?user=%s\\&password=%s", TARGET.attributeWhenReady("datastore.url"), "visitors", "brooklyn", "br00k11n")
```
Is `TARGET` referring to `mysql_server` or `tomcat_server`? I assume the former. What is the case for not using `$brooklyn.component` directly?

sjcorbett commented:
@kiuby88 Thanks. An interesting addition. A few general comments:
- Why use the magic `TARGET` string?  
- The `props.collection` flag feels like implementation details are leaking. Or that the current implementation is conflating two separate issues: how to get attributes from other nodes and where to put that data. 
- Let's come up with a more descriptive name than "Configure"!

@ahgittin Any additional comments?

kiuby88 commented:
@sjcorbett Thanks for your reply.
- I am using `TARGET` instead of `$brooklyn.component` directly in order to start the migration to DSLs to TOSCA syntax. I added this proposal to TOSCA support (Google Doc) as a first approach to support Relationships. However, probably the main lack was I was repeating information (see below). `$brooklyn.component...` contains the same information that `TARGET` node of the relation. So, I thought that maybe it could be a good idea avoiding repeated information using `TARGET` string.

```
relationship_templates:
  database_connection:
    type:   tosca.relationships.configure
    properties:
      prop.collection: java.sysprops
      prop.name: brooklyn.example.db.url
      prop.vale: $brooklyn:formatString("jdbc:%s%s?user=%s\\&password=%s",
         component("mysql_server").attributeWhenReady("datastore.url"), "visitors", "brooklyn", "br00k11n")
```
- You are right. This relation could be complicate to manage. `prop.collection` can represent a Map if `propName` is available. In other case, it represents a list. So, at least one of them must not be empty. I have just added a exception for this case. In any case, I think it is not the best way to manage these alternatives. Probably, I should define three relationships types.
  - Adding a property to a Map.
  - Adding a property to a list.
  - Adding a property as a ConfigKey.
- Probably, taking in account the previous point will be easy to find a name better.

In any case, I'd like to find the way to define all Relationships. In TOSCA support document, I mentioned that we could use TOSCA functions for describing relations or (following Alex's advices) we could use requirement. I have created a [little example](https://gist.github.com/kiuby88/511a6d1fc185e452e08c) which illustrate the second proposal. Moreover, I have added some comments too.

tbouron commented:
@sjcorbett, @kiuby88 

> Why use the magic TARGET string?

That's the whole point of the relationship: not specify the target yourself but getting it automatically from the relationship itself.

> The props.collection flag feels like implementation details are leaking. Or that the current implementation is conflating two separate issues: how to get attributes from other nodes and where to put that data.

We can deal with this in another PR

> Let's come up with a more descriptive name than "Configure"!

What about `brooklyn.relationships.AddProperty` ?

Other than that, LGTM.

robertgmoss commented:
Let's merge now to move forwards. Merging.

sjcorbett commented:
@tbouron On `TARGET`: in the code its uses are simply transformed into `$brooklyn:component(id)`. My comment was that this indirection does nothing that can't be written in the blueprint explicitly.

kiuby88 commented:
Thanks a lot @sjcorbett, @tbouron, @robertgmoss 

@sjcorbett is right. I used a simple transformation and user could use  `$brooklyn:component(id)`. Really, I think this relationship is only a fist (prototype) approach and we should remove it asap. Then, we should decide if we want to use relations based on requirements [example](https://gist.github.com/kiuby88/511a6d1fc185e452e08c) and how we want to use TOSCA functions.
